### PR TITLE
Dense MatrixExponential Transform

### PR DESCRIPTION
**Pull request by stefanwebb**
This is an implementation of the `MatrixExponential` transform for a dense linear operator, y = exp(M)x. The transform was introduced in this recent paper: [The Convolution Exponential and Generalized Sylvester Flows](https://arxiv.org/abs/2006.01910).

The strength of the method is that it allows you to take any linear operation and make it into a bijection with an explicit inverse and tractable log-determinate-Jacobian. I've implemented a dense linear `MatrixExponential` transform here, although the method really shines for structured M on images and graphs (I leave `ConvMatrixExponential` and `GraphMatrixExponential` as well as `GeneralizedSylvester` for future PRs)

### CODE REVIEW

**Comment by stefanwebb** [comment]
@fritzo could we please merge in #2544 before this one?

File: pyro/distributions/transforms/matrix_exponential.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+        y = x.unsqueeze(-1)
+        for idx in range(self.iterations):
+            power_term = torch.matmul(M, power_term) / (idx + 1)
+            y = y + power_term
+
+        return y.squeeze(-1)
+
+    def _trace(self, M):
+        """
+        Calculate the trace of a matrix and is able to do broadcasting over batch
```
Comment (fritzo) [inline pyro/distributions/transforms/matrix_exponential.py:51]: nit: grammar. Calculate -> Calculates

File: pyro/distributions/transforms/matrix_exponential.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    [2] Tim Salimans, Diederik P. Kingma. Weight Normalization: A Simple
+        Reparameterization to Accelerate Training of Deep Neural Networks.
+        [arXiv:1602.07868]
+    [3] Takeru Miyato, Toshiki Kataoka, Masanori Koyama, Yuichi Yoshida. Spectral
+        Normalization for Generative Adversarial Networks. ICLR 2018.
+
+    """
+
+    domain = constraints.real
+    codomain = constraints.real
```
Comment (fritzo) [inline pyro/distributions/transforms/matrix_exponential.py:148]: nit: I guess we could use `constraints.real_vector` here. Do you know if that breaks anything?

Comment (stefanwebb) [reply pyro/distributions/transforms/matrix_exponential.py:148]: It doesn't seem to break anything as far as I'm aware... I should make this change then for the other `event_dim == 1` transforms

File: pyro/distributions/transforms/matrix_exponential.py (hunk 1 above)
Comment (stefanwebb) [reply pyro/distributions/transforms/matrix_exponential.py:51]: Done!

File: pyro/distributions/transforms/matrix_exponential.py (hunk 2 above)
Comment (stefanwebb) [reply pyro/distributions/transforms/matrix_exponential.py:148]: There's no `constraints.real_tensor` though for when `event_dim > 1`...

Comment (stefanwebb) [reply pyro/distributions/transforms/matrix_exponential.py:148]: I've made these changes for all transforms except for `GeneralizedChannelPermute`

**Pull request merged**